pwnlib/encoders/i386/ascii_shellcode.py

```diff
@@ -0,0 +1,300 @@
+        for subtraction in self._calc_subtractions(
+            b'\x00'*int_size, target, vocab):
+            if six.PY2:
+                subtraction = str(subtraction)
+            # sub eax, subtraction
+            result += struct.pack('=c{}s'.format(int_size), b'-', subtraction)
+        result += b'P\\'  # push eax, pop esp
+        # Zero out eax for the unpacking part
+        pos, neg = self._find_negatives(vocab)
+        # and eax, pos; and eax, neg ; (0b00010101 & 0b00101010 = 0b0)
+        result += struct.pack('=cIcI', b'%', pos, b'%', neg)
+        return result
+
+
+    @LocalContext
+    def _find_negatives(self, vocab):
+        r""" Find two bitwise negatives in the vocab so that when they are and-ed the result is 0.
+
+        int_size is taken from the context (context.bits / 8)
+
+        Args:
+            vocab (bytes): Allowed characters
+
+        Returns:
+            Tuple[int, int]: value A, value B
+
+        Raises:
+            ArithmeticError: The allowed character set does not contain
+            two characters that when they are bitwise-anded with eachother
+            they result is 0
+
+        Examples:
+
+            >>> context.update(arch='i386', os='linux')
+            >>> vocab = b'!"#$%&\'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_`abcdefghijklmnopqrstuvwxyz{|}~'
+            >>> a, b = encoders.i386.ascii_shellcode.encode._find_negatives(vocab)
+            >>> a & b
+            0
+        """
+        int_size = context.bits // 8
+        for products in product(*[vocab for _ in range(2)]):
+            if six.PY3:
+                if products[0] & products[1] == 0:
+                    return tuple(int.from_bytes(x.to_bytes(1, 'little')*int_size, 'little') for x in products)
+            elif six.PY2:
+                if six.byte2int(products[0]) & six.byte2int(products[1]) == 0:
+                    return tuple(struct.unpack('=I', x*int_size)[0] for x in products)
+        else:
+            raise ArithmeticError(
+                'Could not find two bitwise negatives in the provided vocab')
+
+
+    @LocalContext
+    def _get_subtractions(self, shellcode, vocab):
+        r""" Covert the sellcode to sub eax and posh eax instructions
+
+        int_size is taken from the context (context.bits / 8)
+
+        Args:
+            shellcode (bytes): The shellcode to pack
+            vocab (bytes): Allowed characters
+
+        Returns:
+            bytes: packed shellcode
+
+        Examples:
+
+            >>> context.update(arch='i386', os='linux')
+            >>> sc = b'ABCDEFGHIGKLMNOPQRSTUVXYZ'
+            >>> vocab = b'!"#$%&\'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_`abcdefghijklmnopqrstuvwxyz{|}~'
+            >>> encoders.i386.ascii_shellcode.encode._get_subtractions(sc, vocab)
+            b'-(!!!-~NNNP-!=;:-f~~~-~~~~P-!!!!-edee-~~~~P-!!!!-eddd-~~~~P-!!!!-egdd-~~~~P-!!!!-eadd-~~~~P-!!!!-eddd-~~~~P'
+        """
+        int_size = context.bits // 8

```

Comment:
We already have a shortcut for this :)
```suggestion
        int_size = context.bytes
```